Sources/KituraNet/HTTP/HTTPServerConfigurationHandler.swift

```diff
@@ -0,0 +1,74 @@
+import NIO
+import NIOHTTP1
+import NIOWebSocket
+import LoggerAPI
+import Foundation
+import Dispatch
+import NIOConcurrencyHelpers
+
+internal class HTTPServerConfigurationHandler: ChannelDuplexHandler {
+    // The HTTPServer instance on which this handler is installed
+    var server: HTTPServer
+    let requestSizeLimit: Int?
+    let connectionLimit: Int?
+    var requestSize: Int = 0
+    var connectionCount = 0
+    typealias InboundIn = ByteBuffer
+    typealias OutboundIn = ByteBuffer
+    typealias InboundOut = ByteBuffer
+    typealias OutboundOut = ByteBuffer
+    
+    public init(for server: HTTPServer) {
+        self.server = server
+//        if let requestSizeLimit = server.serverConfig.requestSizeLimit {
+//            self.requestSizeLimit = requestSizeLimit
+//        }
+//        if let connectionLimit = server.serverConfig.connectionLimit {
+//            self.connectionLimit = connectionLimit
+//        }
+        self.requestSizeLimit = server.serverConfig.requestSizeLimit ?? nil
+        self.connectionLimit = server.serverConfig.connectionLimit ?? nil
+    }
+    public func channelRead(context: ChannelHandlerContext, data: NIOAny) {
+        let data = self.unwrapInboundIn(data)
+        requestSize = requestSize + data.readableBytes
+        if let requestSizeLimit1 = requestSizeLimit {
+            if requestSize > requestSizeLimit1 {
+                let statusDescription = HTTP.statusCodes[HTTPStatusCode.requestTooLong.rawValue] ?? ""
+                var discriptionBuffer = ByteBufferAllocator().buffer(capacity: requestSizeLimit!)
+                discriptionBuffer.writeString(statusDescription)
+                context.writeAndFlush(NIOAny(discriptionBuffer),promise: nil)
+                requestSize = 0
+                context.close(mode: .all, promise: nil)
+                return
+            }
+        }
+        context.fireChannelRead(wrapInboundOut(data))
+    }
+    
+    public func triggerUserOutboundEvent(context: ChannelHandlerContext, event: Any, promise: EventLoopPromise<Void>?) {
+        requestSize = event as? Int ?? 0
```

Comment:
@RudraniW there are a number of issues here:
- the `event` type should be a type private to you
- if the event is of any other type, you should forward it on as you can't expect that you're the only one sending outbound user events (same for inbounds)

But really you're making your life unnecessarily hard. You should be doing the request body size limiting _after_ the HTTP parser. Then the whole limiting is pretty straightforward. As an example, see the handler below:

```swift
final class HTTPRequestBodySizeLimiter: ChannelInboundHandler {
    typealias InboundIn = HTTPServerRequestPart
    typealias InboundOut = HTTPServerRequestPart

    private var bytesReceivedSoFar: Int
    private let maximumAllowedBytes: Int

    init() {
        self.bytesReceivedSoFar = 0
        self.maximumAllowedBytes = 1024 * 1024
    }

    func channelRead(context: ChannelHandlerContext, data: NIOAny) {
        switch self.unwrapInboundIn(data) {
        case .head:
            // at this point in time we must have received 0 bytes
            assert(self.bytesReceivedSoFar == 0)
        case .body(let newBytes):
            guard self.bytesReceivedSoFar + newBytes.readableBytes <= self.maximumAllowedBytes else {
                // for this example, let's just close.
                context.close(promise: nil)
                return
            }
            self.bytesReceivedSoFar += newBytes.readableBytes
        case .end:
            // let's reset the bytes received so far
            self.bytesReceivedSoFar = 0
        }
        // let's forward whatever we have received
        context.fireChannelRead(data)
    }
}
```

The way you would integrate that is

```
channel.pipeline.configureHTTPServerPipeline(...).flatMap {
    channel.pipeline.addHandler(HTTPRequestBodySizeLimiter(...))
}.flatMap {
    channel.pipeline.addHandler(httpHandler)
}
```

Does this make sense?

Reply:
@weissi The objective here was to prevent oversized requests from exhausting the memory on the server.  The reason I suggested the approach of limiting before HTTP header parsing is that a misbehaving client could (deliberately or otherwise) send arbitrary amounts of data as HTTP headers.

I believe that many HTTP servers have a separate limit on the size of the HTTP headers ([Apache](http://httpd.apache.org/docs/2.2/mod/core.html#limitrequestfieldsize) for example), but Kitura does not (at least, not via Kitura-net).  Does Swift-NIO provide a way to limit the size of the headers it accepts?  If so, then imposing separate limits on headers and body would be a superior approach.

Reply:
@djones6 SwiftNIO does have a non-configurable 80kiB limit on the [total size of all headers](https://github.com/apple/swift-nio/blob/master/IntegrationTests/tests_01_http/test_12_headers_too_large.sh).

If you can find anything where SwiftNIO buffers indefinite amounts of anything in memory, please contact us and we will treat it like a security issue.

There's no need to limit anything _in front_ of NIO's HTTP parser or any other parsers.

To make it easy for your own parsers not to grow infinitely in memory, NIO's `ByteToMessageHandler` supports a [`maxBufferSize`](https://apple.github.io/swift-nio/docs/current/NIO/Classes/ByteToMessageHandler.html#/s:3NIO20ByteToMessageHandlerC_17maximumBufferSizeACyxGx_SiSgtcfc) so you as the user can configure any `ByteToMessageDecoder` to not consume more than a fixed amount of buffer space.

Reply:
@RudraniW I think the solution here should probably be two limits: the request header size and the body size (we'd need to implement this in Kitura-net, but that _should_ be a relatively straightforward change on top of the existing prototype).

Reply:
@djones6 you mean two limits for Kitura-net or Kitura-NIO? Because I think for Kitura-NIO you only need to limit what you buffer (the body). Pretty much every web browser/server has a limit and 80 kiB seems large enough for everybody & small enough that it doesn't matter (we took the same value as NodeJS/nginx because it's in http_parser). It doesn't make sense to limit the header block even further because 80 kiB is smaller than most kernel's receive buffers. So even if you limit it in user-space it's in memory already anyway and there's nothing you can do about it. For Kitura-net I'd suggest also just picking an arbitrary value of something like 80kiB, don't think it needs to be configurable tbh.

Reply:
@djones6 btw, there are other things in HTTP/1 that need limiting: the URL and the status description. In NIO we count that towards the header size. So we treat all of those the same:

```
GET /veeeeeeeeerrrrryyyy_long_path_that_might_be_multiple_gigabytes_long [...]
```

```
GET /short/path HTTP/1.1
but: veeeerrryyyyy-long
or: very
many: many
many-many: headers
```

or if we parse responses there can also be

```
HTTP/1.1 200 okaayyyyyyyyy-with-manyyyy-bytes-potentially-multiple-gigabytes
```

So more precisely, NIO's limit is everything until we see `\r\n\r\n` that needs to be maintained  must fit in 80 kiB. Everything after that we stream to you. Oh, and for the trailers the same 80 kiB limit.

Reply:
I think it'll make sense then for Kitura-net to match NIO's 80kb default for headers.  That's also how I would implement it for Kitura-net, which uses CHTTPParser, so we'd impose a header read limit up to the point where the parser indicates that headers are complete (method, path, headers), after which we'd impose the body limit on anything after that.

Reply:
@djones6 cool, if you use `CHTTPParser` (the one from Nods.js), then you'll have the 80kb limit by default (unless you changed it).